File: Enterprise Application Development/bluegather/Services/UsuarioService.cs
using Microsoft.EntityFrameworkCore;
using BlueGather.Dto;
using BlueGather.Models;
using BlueGather.Context;

namespace BlueGather.Services;
public class UsuarioService
{
    private readonly BlueGatherContext _context;

    public UsuarioService(BlueGatherContext context)
    {
        _context = context;
    }

    public async Task<List<UsuarioDto>> FindAll()
    {
        var list = await _context.Usuario
            .Include(x => x.Eventos)
            .ToListAsync();

        return list.Select(entity => ConvertToDto(entity)).ToList();
    }

    public async Task<UsuarioDto> FindById(long id)
    {
        var entity = await FindEntityById(id);
        return ConvertToDto(entity);
    }

    public async Task<UsuarioDto> Create(UsuarioDto newData)
    {
        var entity = await ConvertToEntity(newData);
        _context.Usuario.Add(entity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task<UsuarioDto> Update(long id, UsuarioDto updatedData)
    {
        var entity = await _context.Usuario
            .Include(x => x.Eventos)
            .FirstOrDefaultAsync(x => x.Id == id);

        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}.");
        }

        entity.Eventos.Clear();

        if (updatedData.IdsEventos != null)
        {
            var newEventos = await _context.Evento.Where(e => updatedData.IdsEventos.Contains(e.Id)).ToListAsync();
            foreach (var evento in newEventos)
            {
                entity.Eventos.Add(evento);
            }
        }

        updatedData.Id = entity.Id;
        var updatedEntity = await ConvertToEntity(updatedData);
        _context.Entry(entity).CurrentValues.SetValues(updatedEntity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task Delete(long id)
    {
        var entity = await FindEntityById(id);
        _context.Usuario.Remove(entity);
        await _context.SaveChangesAsync();
    }

    public async Task<UsuarioModel> FindEntityById(long id)
    {
        var entity = await _context.Usuario
            .Include(x => x.Eventos)
            .FirstOrDefaultAsync(x => x.Id == id);
        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}");
        }
        return entity;
    }

    public async Task<List<UsuarioDto>> FindByEventoId(long id)
    {
        var list = await _context.Usuario
            .Where(u => u.Eventos.Any(t => t.Id == id))
            .Include(x => x.Eventos)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    private UsuarioDto ConvertToDto(UsuarioModel entity)
    {
        return new UsuarioDto
        {
            Id = entity.Id,
            Cpf = entity.Cpf,
            Nome = entity.Nome,
            UrlImagem = entity.UrlImagem,
            Email = entity.Email,
            Senha = entity.Senha,
            IdsEventos = entity.Eventos.Select(e => e.Id).ToList()
        };
    }

    private async Task<UsuarioModel> ConvertToEntity(UsuarioDto dto)
    {
        var eventos = new List<EventoModel>();
        if (dto.IdsEventos != null && dto.IdsEventos.Any())
        {
            eventos = await _context.Evento.Where(evento => dto.IdsEventos.Contains(evento.Id)).ToListAsync();
        }

        return new UsuarioModel
        {
            Id = dto.Id ?? 0,
            Cpf = dto.Cpf,
            Nome = dto.Nome,
            UrlImagem = dto.UrlImagem,
            Email = dto.Email,
            Senha = dto.Senha,
            Eventos = eventos
        };
    }
}

File: Digital Business Enablement/bluegather/src/main/java/br/com/fiap/bluegather/service/UsuarioService.java
package br.com.fiap.bluegather.service;

import br.com.fiap.bluegather.dto.UsuarioDTO;
import br.com.fiap.bluegather.model.Usuario;
import br.com.fiap.bluegather.repository.UsuarioRepository;
import br.com.fiap.bluegather.model.Evento;
import br.com.fiap.bluegather.repository.EventoRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class UsuarioService {

    @Autowired
    private UsuarioRepository usuarioRepository;

    @Autowired
    private EventoRepository eventoRepository;

    @Autowired
    PasswordEncoder passwordEncoder;

    public Page<UsuarioDTO> listAll(Pageable pageRequest) {
        Page<Usuario> list = usuarioRepository.findAll(pageRequest);
        return list.map(this::convertToDto);
    }

    public UsuarioDTO findById(Long id) {
        Usuario entity = findEntityById(id);
        return convertToDto(entity);
    }

    @Transactional
    public UsuarioDTO create(UsuarioDTO newData) {
        Usuario entity = convertToEntity(newData);
        Usuario savedEntity = usuarioRepository.save(entity);
        return convertToDto(savedEntity);
    }

    @Transactional
    public UsuarioDTO update(Long id, UsuarioDTO updatedData) {
        findEntityById(id);
        updatedData.setId(id);
        Usuario updatedEntity = convertToEntity(updatedData);    
        Usuario savedEntity = usuarioRepository.save(updatedEntity);
        return convertToDto(savedEntity);
    }
    
    @Transactional
    public void delete(Long id) {
        Usuario entity = findEntityById(id);
        usuarioRepository.delete(entity);
    }

    public Usuario findEntityById(Long id) {
        return usuarioRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "(" + getClass().getSimpleName() + ") - Usuario não encontrado(a) por ID: " + id));
    }

    public UsuarioDTO findByEmail(String email) {
        Usuario entity = usuarioRepository.findByEmail(email)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "(" + getClass().getSimpleName() + ") - Usuario não encontrado(a) por email: " + email));
        return convertToDto(entity);
    }

    public Set<UsuarioDTO> findByEventoId(Long eventoId) {
        Set<Usuario> usuarios = usuarioRepository.findByEventosId(eventoId);
        return usuarios.stream().map(this::convertToDto).collect(Collectors.toSet());
    }
    
    public UsuarioDTO convertToDto(Usuario entity) {
        UsuarioDTO dto = new UsuarioDTO();
        dto.setId(entity.getId());
        dto.setCpf(entity.getCpf());
        dto.setNome(entity.getNome());
        dto.setUrlImagem(entity.getUrlImagem());
        dto.setEmail(entity.getEmail());
        dto.setSenha(null);
        if (entity.getEventos() != null) {
            Set<Long> idsEventos = entity.getEventos().stream()
                    .map(Evento::getId)
                    .collect(Collectors.toSet());
            dto.setIdsEventos(idsEventos);
        }
        return dto;
    }

    public Usuario convertToEntity(UsuarioDTO dto) {
        if (Objects.isNull(dto)) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - UsuarioDTO não pode ser nulo.");
        }
        Usuario entity;
        if (dto.getId() != null) {
            entity = findEntityById(dto.getId());
        } else {
            entity = new Usuario();
        }
        entity.setCpf(dto.getCpf());
        entity.setNome(dto.getNome());
        entity.setUrlImagem((dto.getUrlImagem()));
        entity.setEmail(dto.getEmail());
        entity.setSenha(passwordEncoder.encode(dto.getSenha()));
    
        Set<Evento> newEventos = new LinkedHashSet<>();
        if (dto.getIdsEventos() != null) {
            dto.getIdsEventos().forEach(id -> {
                Evento evento = eventoRepository.findById(id).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Evento não encontrado com ID: " + id));
                newEventos.add(evento);
            });
        }
        entity.setEventos(newEventos);
        return entity;
    }    
}